Project 2 - Cleavage sites and polymorphisms 


INTRODUCTION: 

Scientific Question: How many similar genes are present in the two primate isoform datasets provided by the Daugherty lab? 

Background: 

One of the projects in the daugherty lab is to map primate isoform clevage data to another data set with the same cleavage data across isoforms with population diversity. This way, the lab can acess what genes have population diversity easily without having to reference two data sets constantly. The data is sourced from two memebers of the lab: Brian Tsu (PhD student) and Bryant Cao (Masters student) who are my labmates in the Daugherty lab. 

A little bit of background on viruses. Viruses have proteases that cleave human genes. When human genes are cleaved, some mutatae to avoid further deterioartion and some do not. The first data set has all the clevage site information and the second data set as all the mutataion (diversity) infomration. 
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7150265/
https://pubmed.ncbi.nlm.nih.gov/3444396/

Hypothesis: If there are clevage sites for a gene, then it must have population diversity. 

Description: 
There are 95,265 genes with clevage site positions in the data set "counts_tog_AGM_ORFeome_mapping" which I will be referring to as data set 1. There 1199 genes with population diversity in the data set "filtered_pascaf_codon_level_data" which I will be referring to as data set 2. There are repititions of some gene in data set 1. First, I imported the data sets as a CSV file. My lab mates taught me how to do that! Next, I found the stop site of the cleavage site by using python to add 7 to the start sites because its an 8-mer using numpy. Next, I did some mini tasks like adding 8 new columns to data set 1 and moving the column order just for cleanliness. Next, I found out the number of similar genes in both datasets. Then I removed the duplicate genes and found the real number of common genes. Lastly, I exported the data to an excel sheet using excel writer. The hypothesis I made was wrong. If every gene with a clevage site had human diversity, then I would not have to do this project! Not all genes evolve after the presence of a clevage site. Which is why we see only 737 genes at the end that have human diversity data. 






LOADING IN PACKAGES: 

Pandas: Pandas is a very popular python package used to manipulate large data sets. It is very rapid and easy once people get used to it. It is useful in doing real-world data analysis and is the most powerful open-source manipulation tool available in any language. Pandas can be used to manipulate excel sheet data, tabular data, ordered or unordered data, statistical data sets etc. The two main data structures of pandas, series and data frame, handle the vast majority of typical uses.
I used Pandas to import my data sets as csv files. Pandas helped me to manipulate data in the csv file column by column when I was checking if my code worked. I also used pandas to handle missing data, delete an dinsert columns, align data, index data, subsetting data, and joining data sets. 

Here are some additional uses/description of pandas:
Easy handling of missing data: represented as NaN

Size mutability: columns can be inserted and deleted from DataFrame 

Automatic and explicit data alignment: objects can be explicitly aligned to a set of labels

Make it easy to convert differently-indexed data in other Python and NumPy data structures into DataFrame objects

Intelligent label - fancy indexing, subsetting of large data sets, merging and joining data sets

Robust IO tools for loading data from flat files - (CSV and delimited), Excel files etc.  


#https://pandas.pydata.org/pandas-docs/stable/getting_started/overview.html
https://pypi.org/project/pandas/


NumPy: NumPy stands for Numerical python. It is a Python library tool utilized to work with arrays predominantly. It has functions for working in domains like linear algebra, matrices etc. It is also an open-source project like pandas. In fact, Pandas was built on top of NumPy. and you can use it freely. NumPy is faster than lists. NumPy arrays are usually stored at one continuous place in memory making the processes of accessing and manipulation very effective. NumPy is a mostly written in python but can also be used in C and C++. I used numpy to turn the rowns and columns into lists so its easier for me to delete and add data points.
Here are some additional uses/description of numpy: 

POWERFUL N-DIMENSIONAL ARRAYS -  the NumPy vectorization is fast and versatile. Indexing and broadcasting concepts are the de-facto standards of array computing today in the coding world.

NUMERICAL COMPUTING TOOLS - NumPy offers comprehensive mathematical functions, random number generators, linear algebra routines, Fourier transforms etc. 

INTEROPERABLE - NumPy supports a range of computing platforms and plays well with distributed sparse array libraries.

EASY TO USE - NumPy’s high level syntax makes it accessible and effective for programmers from any background or experience level. 

#https://www.w3schools.com/python/numpy/numpy_intro.asp#:~:text=NumPy%20is%20a%20Python%20library,NumPy%20stands%20for%20Numerical%20Python.
https://numpy.org/





import pandas as pd
import numpy as np

In [ ]:
import pandas as pd 
import numpy as np

#PERFORMING BIOINFORMATICS ANALYSIS: 

#Here, I could not really plot the data. Instead I've decided to use ExcelWriter to export the data of common isoform genes to an excel sheet. This way I can acess it easily and import it as a csv file if I need to maipulate it in the future. 

#ExcelWriter: Is a commonly used tool for writing dataframe objects into excel sheets. 
#ex: with ExcelWriter('path_to_file.xlsx') as writer:
    df.to_excel(writer)

#https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.ExcelWriter.html

#I know there are other ways to visualize data, but in this case I believe that this is the best option. We have the number of genes in common and I can visualize it in the pandas excel output.


In the cell below, I am adding 7 to the start positions to find the end position of the cleavage site in a 8-mer. I try doing this for one row first and then do it to all the rows with different genes. 

In [12]:
import pandas as pd 

df = pd.read_csv('counts_tog_AGM_ORFeome_mapping.csv')

#Take start col, generate stop positions
a = df['start']#same as df.start -- here we are just previewing values from the df col named 'start'
#print(a) #quick diagnostic print
df['stop_pos']=df['start']+7 #add 7 to every single element in df['start'] and save to new col named 'stop_pos'
#print(df['stop_pos'])

#Goal #1: Try to move stop_pos col next to start, for cleanliness, but very optional
#Other side goals: renaming cols

#Goal #2: Create now 8 new columns, you may keep them blank/with 0's OR fill them immediately upon creation (as we did with stop_pos) -- If fill with real values, see goal #3

#Goal #3: Map IDs of same type between two csv files, then extract associated values fro one of the csv files
#Use NM_IDs from the cleavage dataframe to match with NM_IDs from Bryant's table and then subsequently pull associated, desired data
#A) Find all entries in Pascaf table with matching ID


Stop positions for all the genes 

In [40]:
# Generating stop values. For a detailed version, look at Counts tog.ipynb 
import pandas as pd 
#making the ORF table a dataframe  
df.ORF = pd.read_csv('counts_tog_AGM_ORFeome_mapping.csv')
#defining the variable a as the start value column
a = df.ORF['start']
#defining the stop value column as the start+7 
df.ORF['stop']=df.ORF['start']+7
#printing the stop values 
print(df.ORF['stop'])
#Adding these stop values to the excel sheet. Can change the excel sheet name if I wanted to add ot to a different one 
df.ORF['stop'].to_excel('counts_tog_AGM_ORFeome_mapping.xlsx')
len(df.ORF['stop'])

0        230
1        234
2        991
3        352
4        514
        ... 
95261    375
95262    804
95263     21
95264    225
95265    359
Name: stop, Length: 95266, dtype: int64


95266

In the cell below, I move the stop column next to the start column for cleanliness. The output is seen as a list. 

In [35]:
# moving stop col next to start

import pandas as pd 

df.ORF = pd.read_csv('counts_tog_AGM_ORFeome_mapping.csv')
titles = list(df.ORF.columns)
print(titles)

new_columns=(['_Counts_','Gene_Sym','Ref_SEQ','start','stop','Nuc_start','Num_missing','Num_present','Hit_Ratio','Total_AA_Length','Trimer_motif','Unique_Seqs','Concat_Aln','Gene_Desc','Two_Fold_Up','Five_Fold_Up','Two_Fold_Down','Five_Fold_Down','PC1','Best_pval_org','Best_pval','Human_pval','AA_String','Toggle_Num','Gene_Tog_Num','AA_String_Map','Pos_Sel_Res','P4','P3','P2','P1','P1p','P2p','P3p','P4p','AGM_pval','ORF_ID','Resource_Plate','Resource_Position','ORF_Nuc_Length','Nuc_start'])










['_Counts_', 'Gene_Sym', 'Ref_SEQ', 'start', 'Nuc_start', 'Num_missing', 'Num_present', 'Hit_Ratio', 'Total_AA_Length', 'Trimer_motif', 'Unique_Seqs', 'Concat_Aln', 'Gene_Desc', 'Two_Fold_Up', 'Five_Fold_Up', 'Two_Fold_Down', 'Five_Fold_Down', 'PC1', 'Best_pval_org', 'Best_pval', 'Human_pval', 'AA_String', 'Toggle_Num', 'Gene_Tog_Num', 'AA_String_Map', 'Pos_Sel_Res', 'P4', 'P3', 'P2', 'P1', 'P1p', 'P2p', 'P3p', 'P4p', 'AGM_pval', 'ORF_ID', 'Resource_Plate', 'Resource_Position', 'ORF_Nuc_Length', 'Unnamed: 39', 'stop']


Next, I added 8 new columns to the data set. Each column refers to each site in the 8-mer. If there is polymorphism in each site, we can indicate it in th enewly formed columns. 

In [50]:
#Adding 8 new columns 

import numpy as np 
#adds 95266 zeros in the column 'zero'
zero = np.zeros( (95266))
df.ORF['zero']=zero
#adds 95266 zeros in the column 'one'
one=np.zeros( (95266))
df.ORF['one']=one
#adds 95266 zeros in the column 'two'
two=np.zeros( (95266))
df.ORF['two']=two
#adds 95266 zeros in the column 'three'
three=np.zeros( (95266))
df.ORF['three']=three
#adds 95266 zeros in the column 'four'
four=np.zeros( (95266))
df.ORF['four']=four
#adds 95266 zeros in the column 'five'
five=np.zeros( (95266))
df.ORF['five']=five
#adds 95266 zeros in the column 'six'
six=np.zeros( (95266))
df.ORF['six']=six
#adds 95266 zeros in the column 'seven'
seven=np.zeros( (95266))
df.ORF['seven']=seven

df.ORF
with pd.ExcelWriter('Counts.xlsx') as writer:
    df.ORF.to_excel(writer)




In the cell below, I cross check two data sets to find out the common gene names in both of the data set sand print them out. The ouput shows the common gene IDs. 

In [58]:
#create a list of the two ref_sequene columns 
column_names = ["Ref_S",'RefSeq']
#create a dataframe, specify the csv file name and assign names to the column_names list you created above
df = pd.read_csv('Ref_Sequences.csv', names=column_names)
#now change the dataframes to list
Ref_S = df.Ref_S.to_list()
RefSeq = df.RefSeq.to_list()
#create a new list to append the matched IDs to
new_list=[]
#For every num in list1, print num1 only is num is equal to num1. num and num1 are referring to each element in the list 
for num in Ref_S:
    for num1 in RefSeq:
        if num == num1:
            new_list.append(num)
print(new_list)


, 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536', 'NM_004536

In [59]:
#There are duplicates in the clevage table. So, the matched IDs are more than th etotal number of Ref-Seq Ids in the pascaf table. The duplicates are deleted in the cell below. 
len(new_list)

7684

In [61]:
#Removed the duplicates from the clevage table
final_answer=[]
#The code says- add the number to the final_answer above unless the number is already present in the final_answer list. This removes the duplicates in the new_list. So, if the clevage table has five 3's and pascaf table has one 3, the new_list will have five 3's and the final answer will have only one 3. The duplicates are being removed
for i in new_list:
    if i not in final_answer:
        final_answer.append(i)

len(final_answer)
print(final_answer)

['NM_024763', 'NM_014053', 'NM_015585', 'NM_001136219', 'NM_001005167', 'NM_001105569', 'NM_003433', 'NM_001166305', 'NM_000448', 'NM_001195831', 'NM_001145460', 'NM_004224', 'NM_001039503', 'NM_004688', 'NM_001282478', 'NM_001145434', 'NM_001291978', 'NM_001169', 'NM_015210', 'NM_023014', 'NM_001040462', 'NM_174952', 'NM_002278', 'NM_014272', 'NM_001044369', 'NM_006065', 'NM_002864', 'NM_145053', 'NM_001242894', 'NM_014757', 'NM_017681', 'NM_152658', 'NM_138477', 'NM_181645', 'NM_001134224', 'NM_032816', 'NM_001099695', 'NM_001302652', 'NM_024164', 'NM_001282866', 'NM_001177599', 'NM_001164425', 'NM_021268', 'NM_001145121', 'NM_001040025', 'NM_004388', 'NM_152329', 'NM_001013734', 'NM_006440', 'NM_003679', 'NM_152657', 'NM_001207052', 'NM_001037163', 'NM_001127178', 'NM_005385', 'NM_001007532', 'NM_001127610', 'NM_001282468', 'NM_001190482', 'NM_001004467', 'NM_001184', 'NM_015272', 'NM_017662', 'NM_015307', 'NM_031886', 'NM_002425', 'NM_019121', 'NM_031282', 'NM_080833', 'NM_003211',

In [65]:
import pandas as pd 
#Converting the list to a datframe. Name a new dataframe and define it as pd. DataFrame(), with the name of the list and the name of the column
df.Matched_IDs = pd.DataFrame(final_answer,columns=['Matched_IDs'])
print(df.Matched_IDs)

0         NM_024763
1         NM_014053
2         NM_015585
3      NM_001136219
4      NM_001005167
           ...     
732       NM_022436
733       NM_024046
734    NM_001014342
735       NM_148901
736    NM_001102445
Name: Matched_IDs, Length: 737, dtype: object


In [69]:
df.Matched_IDs
with pd.ExcelWriter('Matched_IDs.xlsx') as writer:
    df.Matched_IDs.to_excel(writer)


Analyzing the Data: 

The scientific question is answered by the code above. Out of 95,266 genes, there are 7684 genes in common (including the duplicates) and 737 when the duplicates are removed. The data set with the polymorphism information indicated that not all the genes with cleavage sites in the dataset 1 have mutations. This proves that having a clevage site does not mean that there will be polymorphism in the gene. Some sites evolve to avoid clevage by viral proteases. However, some reamain the same. Hence, the hypothesis is proved wrong. Having a cleavage site in a gene does not mean that it will have population diversity. 